Sources/PackageCollectionsSigning/Key/Key+RSA.swift

```diff
@@ -0,0 +1,162 @@
+        let data = try KeyUtilities.stripHeaderAndFooter(pem: pem)
+
+        let options: [String: Any] = [
+            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
+            kSecAttrKeyClass as String: kSecAttrKeyClassPrivate,
+        ]
+
+        var error: Unmanaged<CFError>?
+        guard let key = SecKeyCreateWithData(data as CFData,
+                                             options as CFDictionary,
+                                             &error) else {
+            throw error.map { $0.takeRetainedValue() as Error } ?? KeyError.initializationFailure
+        }
+
+        self.underlying = key
+    }
+}
+
+struct CoreRSAPublicKey: PublicKey {
+    let underlying: SecKey
+
+    /// `data` should be in PKCS #1 format
+    init(data: Data) throws {
+        let options: [String: Any] = [
+            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
+            kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
+        ]
+
+        var error: Unmanaged<CFError>?
+        guard let key = SecKeyCreateWithData(data as CFData,
+                                             options as CFDictionary,
+                                             &error) else {
+            throw error.map { $0.takeRetainedValue() as Error } ?? KeyError.initializationFailure
+        }
+
+        self.underlying = key
+    }
+
+    init<Data>(pem data: Data) throws where Data: DataProtocol {
+        let pem = String(decoding: data, as: UTF8.self)
+        let data = try KeyUtilities.stripHeaderAndFooter(pem: pem)
+        try self.init(data: data)
+    }
+}
+
+// MARK: - RSA key implementations using BoringSSL
+
+// Reference: https://github.com/vapor/jwt-kit/blob/master/Sources/JWTKit/RSA/RSAKey.swift
+
+#else
+final class BoringSSLRSAPrivateKey: PrivateKey, BoringSSLKey {
+    let underlying: UnsafeMutablePointer<CCryptoBoringSSL.RSA>
+    let algorithm: OpaquePointer = CCryptoBoringSSL_EVP_sha256()
+
+    deinit {
+        CCryptoBoringSSL_RSA_free(self.underlying)
+    }
+
+    init<Data>(pem data: Data) throws where Data: DataProtocol {
+        let key = try Self.load(pem: data) { bio in
+            CCryptoBoringSSL_PEM_read_bio_PrivateKey(bio, nil, nil, nil)
+        }
+        defer { CCryptoBoringSSL_EVP_PKEY_free(key) }
+
+        guard let pointer = CCryptoBoringSSL_EVP_PKEY_get1_RSA(key) else {
+            throw BoringSSLKeyError.rsaConversionFailure
+        }
+
+        self.underlying = pointer
+    }
+}
+
+final class BoringSSLRSAPublicKey: PublicKey, BoringSSLKey {
+    let underlying: UnsafeMutablePointer<CCryptoBoringSSL.RSA>
+    let algorithm: OpaquePointer = CCryptoBoringSSL_EVP_sha256()
+
+    deinit {
+        CCryptoBoringSSL_RSA_free(self.underlying)
+    }
+
+    /// `data` should be in the PKCS #1 format
+    init(data: Data) throws {
+        var bytes: UnsafePointer<UInt8>? = try data.withUnsafeBytes {
+            guard let bytes = $0.bindMemory(to: UInt8.self).baseAddress else {
```

Comment:
It's also not safe to bind the memory this way either. You're using `copyBytes` in other places, why aren't you doing that here?

Reply:
amend https://github.com/apple/swift-package-manager/pull/3238/commits/6a0489b50be0d4143959ca7e20bbb8d750323b76